File: ufa/services/analytical-moose-foobar/app/apis/foo/consumption/foo-cube-aggregations-api.ts
import { ConsumptionApi } from "@514labs/moose-lib";
import { FooPipeline } from "../../../index";
import {
  GetFooCubeAggregationsParams,
  GetFooCubeAggregationsResponse,
} from "@workspace/models/foo";

export const fooCubeAggregationsApi = new ConsumptionApi<
  GetFooCubeAggregationsParams,
  GetFooCubeAggregationsResponse
>(
  "foo-cube-aggregations",
  async (
    {
      months = 6,
      status,
      tag,
      priority,
      limit = 20,
      offset = 0,
      sortBy,
      sortOrder = "ASC",
    }: GetFooCubeAggregationsParams,
    { client, sql }
  ): Promise<GetFooCubeAggregationsResponse> => {
    const startTime = Date.now();

    const endDate = new Date();
    const startDate = new Date();
    startDate.setMonth(
      startDate.getMonth() - Math.max(1, Math.min(months, 36))
    );

    const startDateStr = startDate.toISOString().split("T")[0];
    const endDateStr = endDate.toISOString().split("T")[0];

    // Build optional WHERE fragments
    const statusFilter = status ? sql`AND status = ${status}` : sql``;
    const priorityFilter =
      typeof priority === "number" ? sql`AND priority = ${priority}` : sql``;

    const limited = Math.max(1, Math.min(limit, 200));
    const pagedOffset = Math.max(0, offset);

    // Map sort column safely
    const sortColumn = (() => {
      switch (sortBy) {
        case "month":
          return sql`month`;
        case "status":
          return sql`status`;
        case "tag":
          return sql`tag`;
        case "priority":
          return sql`priority`;
        case "n":
          return sql`n`;
        case "avgScore":
          return sql`avgScore`;
        case "p50":
          return sql`p50`;
        case "p90":
          return sql`p90`;
        default:
          return sql`month, status, tag, priority`;
      }
    })();
    const sortDir = sql([sortOrder.toUpperCase() === "DESC" ? "DESC" : "ASC"]);

    const query = sql`
      WITH exploded AS (
        SELECT
          toStartOfMonth(created_at) AS month,
          status,
          arrayJoin(tags) AS tag,
          priority,
          score
        FROM ${FooPipeline.table!}
        WHERE toDate(created_at) >= toDate(${startDateStr})
          AND toDate(created_at) <= toDate(${endDateStr})
          AND score IS NOT NULL
          AND cdc_operation != 'DELETE'
          ${statusFilter}
          ${priorityFilter}
      ),
      aggregated AS (
        SELECT
          CASE 
            WHEN month IS NULL THEN NULL 
            ELSE formatDateTime(month, '%Y-%m-01') 
          END AS month,
          toNullable(status) AS status,
          toNullable(tag) AS tag,
          toNullable(priority) AS priority,
          count() AS n,
          avg(score) AS avgScore,
          quantileTDigest(0.5)(score) AS p50,
          quantileTDigest(0.9)(score) AS p90
        FROM exploded
        ${tag ? sql`WHERE tag = ${tag}` : sql``}
        GROUP BY CUBE(month, status, tag, priority)
        HAVING month IS NOT NULL AND status IS NOT NULL AND tag IS NOT NULL AND priority IS NOT NULL
      )
      SELECT
        *,
        COUNT() OVER() AS total
      FROM aggregated
      ORDER BY ${sortColumn} ${sortDir}
      LIMIT ${limited} OFFSET ${pagedOffset}
    `;

    const resultSet = await client.query.execute<{
      month: string | null;
      status: string | null;
      tag: string | null;
      priority: number | null;
      n: number;
      avgScore: number;
      p50: number;
      p90: number;
      total: number;
    }>(query);

    const rows = (await resultSet.json()) as {
      month: string | null;
      status: string | null;
      tag: string | null;
      priority: number | null;
      n: number;
      avgScore: number;
      p50: number;
      p90: number;
      total: number;
    }[];

    const queryTime = Date.now() - startTime;
    const total = rows.length > 0 ? Number(rows[0].total) : 0;

    return {
      data: rows.map((r) => ({
        month: r.month,
        status: r.status,
        tag: r.tag,
        priority: r.priority === null ? null : Number(r.priority),
        n: Number(r.n),
        avgScore: Number(r.avgScore),
        p50: Number(r.p50),
        p90: Number(r.p90),
      })),
      queryTime,
      pagination: {
        limit: limited,
        offset: pagedOffset,
        total,
      },
    };
  }
);

File: ufa-lite/services/analytical-moose-foobar-py/app/apis/foo/consumption/foo_cube_aggregations_api.py
from moose_lib import Api
from typing import Optional, List
from pydantic import BaseModel
from app.external_models import foo_model
import time
from datetime import datetime, timedelta


class GetFooCubeAggregationsParams(BaseModel):
    months: Optional[int] = 6
    status: Optional[str] = None
    tag: Optional[str] = None
    priority: Optional[int] = None
    limit: Optional[int] = 20
    offset: Optional[int] = 0
    sort_by: Optional[str] = None
    sort_order: Optional[str] = "ASC"


class FooCubeAggregationRow(BaseModel):
    month: Optional[str] = None
    status: Optional[str] = None
    tag: Optional[str] = None
    priority: Optional[int] = None
    n: int
    avgScore: float
    p50: float
    p90: float


class GetFooCubeAggregationsResponse(BaseModel):
    data: List[FooCubeAggregationRow]
    queryTime: int
    pagination: dict


def foo_cube_aggregations_api_handler(
    context,
    params: GetFooCubeAggregationsParams
) -> GetFooCubeAggregationsResponse:
    """
    Cube aggregations API for foo data with monthly grouping and statistical analysis
    """
    start_time = time.time()

    # Calculate date range
    end_date = datetime.now()
    start_date = end_date - timedelta(days=max(1, min(params.months, 36)) * 30)

    start_date_str = start_date.strftime("%Y-%m-%d")
    end_date_str = end_date.strftime("%Y-%m-%d")

    # Build optional WHERE fragments
    where_clauses = [
        f"toDate(created_at) >= toDate('{start_date_str}')",
        f"toDate(created_at) <= toDate('{end_date_str}')",
        "score IS NOT NULL"
    ]

    if params.status:
        where_clauses.append(f"status = '{params.status}'")

    if params.priority is not None:
        where_clauses.append(f"priority = {params.priority}")

    limited = max(1, min(params.limit, 200))
    paged_offset = max(0, params.offset)

    # Map sort column safely
    sort_column = {
        "month": "month",
        "status": "status",
        "tag": "tag",
        "priority": "priority",
        "n": "n",
        "avgScore": "avgScore",
        "avg_score": "avgScore",
        "p50": "p50",
        "p90": "p90"
    }.get(params.sort_by, "month, status, tag, priority")

    sort_dir = "DESC" if params.sort_order.upper() == "DESC" else "ASC"

    # Build the main query
    where_clause = " AND ".join(where_clauses)
    having_clause = f"HAVING tag = '{params.tag}'" if params.tag else ""

    query = f"""
        SELECT
            formatDateTime(toStartOfMonth(created_at), '%Y-%m-01') AS month,
            status,
            arrayJoin(tags) AS tag,
            priority,
            count() AS n,
            avg(score) AS avgScore,
            quantileTDigest(0.5)(toFloat64(score)) AS p50,
            quantileTDigest(0.9)(toFloat64(score)) AS p90,
            COUNT() OVER() AS total
        FROM foo
        WHERE {where_clause}
        GROUP BY month, status, tag, priority
        {having_clause}
        ORDER BY {sort_column} {sort_dir}
        LIMIT {limited} OFFSET {paged_offset}
    """

    results = context.query(query, {})

    query_time = int((time.time() - start_time) * 1000)
    total = results[0]["total"] if results else 0

    # Convert results to response format
    data = [
        FooCubeAggregationRow(
            month=row["month"],
            status=row["status"],
            tag=row["tag"],
            priority=int(row["priority"]) if row["priority"] is not None else None,
            n=int(row["n"]),
            avgScore=float(row["avgScore"]),
            p50=float(row["p50"]),
            p90=float(row["p90"])
        )
        for row in results
    ]

    return GetFooCubeAggregationsResponse(
        data=data,
        queryTime=query_time,
        pagination={
            "limit": limited,
            "offset": paged_offset,
            "total": total
        }
    )


# Create the API instance
foo_cube_aggregations_api = Api[GetFooCubeAggregationsParams, GetFooCubeAggregationsResponse](
    name="foo-cube-aggregations",
    query_function=foo_cube_aggregations_api_handler
)